articles/101_build_tool.md

```diff
@@ -0,0 +1,213 @@
+
+#### Build ROS 1 workspaces
+
+The tool needs to be able to build ROS 1 workspaces which can already be built using `catkin_make_isolated` / `catkin_tools`.
+
+#### Build ROS 2 workspaces
+
+The tool needs to be able to build ROS 2 workspaces which can already be built using `ament_tools`.
+
+#### Build Gazebo including dependencies
+
+After cloning the repositories containing Gazebo and all its dependencies (excluding system packages) the tool needs to be able to build the set of packages.
+Meta information not inferable from the sources can be provided externally without adding or modifying any files in the workspace.
+After the build a single file can be sourced / invoked to setup the environment to use Gazebo (e.g. `GAZEBO_MODEL_PATH`).
+
+## Software Criteria
+
+The tool aims to support a variety of build systems, use cases, and platforms.
+The above mentioned ones are mainly driven by the needs in the ROS ecosystem but the tool should also be usable outside the ROS ecosystem (e.g. for Gazebo).
+Therefore it should be designed in a way which enables extending its functionality.
+
+Assuming that the tool will be implemented in Python (since that is the case for existing tools) the entry point mechanism provides a convenient way to make the software extensible.
+Extensions don't even have to be integrated into the Python package containing the core logic of the build tool but can easily be provided by additional Python packages.
+This approach will not only foster a modular design and promote clear interfaces but enable external contributions without requiring them to be integrated in a single monolithic package.
+
+Several well known software principles apply:
+
+- Separation of concerns
+- Single Responsibility principle
+- Principle of Least Knowledge
+- Don’t repeat yourself
+- Keep it stupid simple
+- "Not paying for what you don't use"
+
+### Extension Points
+
+The following items are possible extension points to provide custom functionality:
+
+- contribute `verbs` to the command line tool (e.g. `build`, `test`)
+- contribute command line options for specific features (e.g. `build`, `test`)
+- discovery of packages (e.g. recursively crawling a workspace)
+- identification of packages and their meta information (e.g. from a `package.xml` file)
+- process a package (e.g. build a CMake package, test a Python package)
+- execution model (e.g. sequential processing, parallel processing)
+- output handling (e.g. console output, logfiles, status messages, notifications)
+- setup the environment (e.g. `sh`, `bash`, `bat`)
+- completion (e.g. `bash`, `Powershell`)
+
+## Alternative Approaches
+
+There are two different approaches possible to reach the goal of a universal build tool.
+
+- Incrementally evolve one of the existing tools to satisfy the described goals.
+- Implement a new build tool "from scratch" (while existing building blocks can still be "copied" from the existing solutions).
```

Comment:
Also, are there any other tools out there which provide some of this functionality, particularly the ones about being extensible?  Some other ecosystem of existing tooling could help us get over the initial startup cost of something new.